common/src/main/java/com/wynntils/screens/itemfilter/ItemFilterScreen.java

```diff
@@ -0,0 +1,1279 @@
+
+                Collections.swap(presets, presetIndex, indexToSwap);
+
+                Services.ItemFilter.presets.store(presets);
+                Services.ItemFilter.presets.touched();
+
+                presets = Services.ItemFilter.presets.get();
+
+                populatePresets();
+            } else { // Select the preset
+                itemSearchWidget.setTextBoxInput(presets.get(presetIndex).b());
+            }
+        } else if (button == GLFW.GLFW_MOUSE_BUTTON_RIGHT) {
+            if (KeyboardUtils.isShiftDown()) { // Shift the preset down
+                int indexToSwap = presetIndex == presets.size() - 1 ? 0 : presetIndex + 1;
+
+                Collections.swap(presets, presetIndex, indexToSwap);
+
+                Services.ItemFilter.presets.store(presets);
+                Services.ItemFilter.presets.touched();
+
+                presets = Services.ItemFilter.presets.get();
+
+                populatePresets();
+            } else if (KeyboardUtils.isControlDown()) { // Delete the preset
+                presets.remove(presetIndex);
+
+                Services.ItemFilter.presets.store(presets);
+                Services.ItemFilter.presets.touched();
+
+                presets = Services.ItemFilter.presets.get();
+
+                presetsScrollOffset = Math.max(presetsScrollOffset - 1, 0);
+
+                populatePresets();
+
+                if (presets.size() <= MAX_PRESETS) {
+                    nextPresetButton.visible = false;
+                    previousPresetButton.visible = false;
+                }
+            }
+        }
+    }
+
+    private void savePreset() {
+        presets.add(new Pair<>(presetNameInput.getTextBoxInput(), itemSearchWidget.getTextBoxInput()));
+
+        Services.ItemFilter.presets.store(presets);
+        Services.ItemFilter.presets.touched();
+
+        presetNameInput.setTextBoxInput("");
+
+        if (presets.size() > MAX_PRESETS) {
+            nextPresetButton.visible = true;
+            previousPresetButton.visible = true;
+        }
+
+        populatePresets();
+    }
+
+    private void checkSaveStatus() {
+        if (presetNameInput == null || itemSearchWidget == null) return;
+
+        savePresetButton.active = !presetNameInput.getTextBoxInput().isEmpty()
+                && !itemSearchWidget.getTextBoxInput().isEmpty();
+    }
+
+    private void toggleSortMode() {
+        sortMode = !sortMode;
+
+        itemNameInput.visible = !sortMode;
+
+        if (sortMode) {
+            if (valueWidget != null) {
+                selectedProvider = null;
+                this.removeWidget(valueWidget);
+            }
+
+            sortScrollOffset = 0;
+
+            populateSorts();
+        } else {
+            for (AbstractWidget widget : sortButtons) {
+                this.removeWidget(widget);
+            }
+
+            sortButtons = new ArrayList<>();
+        }
+    }
+
+    private void parseFilters(String input) {
```

Comment:
I would assume this is the hacky part you mentioned before. At first sight, my first idea/question is why not just use `ItemFilterService#createSearchQuery`? You get back a parsed object, from that, you can easily get all providers/sorts/whatever you need. Would this work?

Reply:
I overall think  that the widget's `updateValues` do too much parsing, I assume you know this as well. I wonder if you would still need this parsing if'd done the suggestion above, since `ItemSearchQuery` has the StatFilters too, perhaps it could be passed to the widget to make parsing easier since you would only have to decide how you represent the filter in the widget itself, without needing to parse it.

Reply:
I think `parseFilters` is fine, it's the `updateValues` I'm not a fan of especially on the `IntegerValueWidget`. I'll look into using the services method, maybe that will be better

Reply:
I tried using this but was unable to get it to work for filters. Sorts and plain queries were fine but after removing the parsing from the constructors and always calling `updateValues` I'm happier with how it works. The `ignoreUpdate` field in the widgets still bothers me a bit but I don't see how it can be avoided, unless I add a `setTextBoxInputWithoutConsuming`..

Reply:
I'll try to make a patch to make this work.